Merge master to code_review

ConCEAT opened the pull request:
Hope this one works

ConCEAT committed:
add comics endpoints, add dotenv

Add endpoints to:
* /GET /comics/current
* /GET /comics/many
* /GET /comics/<comicID>

Add dotenv for future local variables and config

ConCEAT committed:
add download endpoint

ConCEAT committed:
separate non-endpoint functions from endpoint functions

ConCEAT committed:
delete separated functions from main

ConCEAT committed:
add types and docs to functions.py

ConCEAT committed:
Change error output

Change error respresentation from error.msg to full error json
Add for endpoints:
*/comics/many
*/comics/download
If any comic_ids raises error,
endpoint returns only error representation

ConCEAT committed:
change return None to return empty JSON in download endpoint

papisz commented on functions.py:
```diff
@@ -0,0 +1,44 @@
+from typing import Dict
+
+
+def getFilenames(path: str) -> list:
```
Python style guide recommends using `snake_case` as function names :) 
https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

It's best to enable these checks in your code editor

papisz commented on functions.py:
```diff
@@ -0,0 +1,44 @@
+    """
+    fileExtension = url.split('.')[-1]
+    image = requests.get(url).content
+    with open(f'{path}{filename}.{fileExtension}', 'wb') as newImage:
```
the safer way to do it is [os.path.join](https://docs.python.org/3/library/os.path.html#os.path.join) - this works on all operating systems and handles the `/` or `\` for you :)

papisz commented on main.py:
```diff
@@ -0,0 +1,77 @@
+@app.get("/comics/{comicId}")
+def getComicByID(comicId: int, response: Response):
+    try:
+        comicData = functions.loadJSON(f'https://xkcd.com/{comicId}/info.0.json')
```
usually it's best to put the first part of the URL in an environmental variable - for example:

```
f'{host}/{comicId}/info.0.json'
```

In this case it doesn't matter so much, but sometimes URLs change and it's best to be prepared for it.

papisz commented on main.py:
```diff
@@ -0,0 +1,77 @@
+    except urllib.error.HTTPError as error:
+        response.status_code = error.code
+        return error
+    return functions.processComicData(comicData)
```
is it a good idea to have a module named `functions`? what happens when your application grows?

try to think about separating the logic a little bit more :). Maybe add some classes that represent objects, like a comic for example?

papisz commented on main.py:
```diff
@@ -0,0 +1,77 @@
+def getManyComics(response: Response, comic_ids: List[int] = Query(...)):
+    output = []
+    history = []
+    for comicId in comic_ids:
```
there are some structures which take care of uniqueness for you - for example a dictionary and a set guarantee that all keys/elements are unique. Try using one of them here.

papisz commented on main.py:
```diff
@@ -0,0 +1,77 @@
+    output = []
+    history = []
+    for comicId in comic_ids:
+        if comicId in history:
```
this looks simple and nice, BUT there's one problem with it - when `history` is a list, what _really_ happens is: the interpreter needs to iterate through all elements until it finds `comicId`.

If `history` was a set or a list, then this operation would be much cheaper. 

Look at this example:

```
$ python -m timeit -u usec -n 10000 -s "a = list(range(100))" "99 in a"
10000 loops, best of 5: 1.13 usec per loop
 $ python -m timeit -u usec -n 10000 -s "a = set(range(100))" "99 in a"
10000 loops, best of 5: 0.0278 usec per loop
```

ConCEAT committed:
switch names to snake_case, use os.path.join

ConCEAT committed:
Store host adress in an environmental variable

ConCEAT committed:
Refactor: optimize request args uniqueness

Use dict to create and store unique hashtable for locally
stored images and already maintained images.

ConCEAT committed:
Refactor: rename and reorganise logic into classes

ConCEAT committed:
Refactor: minor formatting changes
